src/main/scala/model/Bunny.scala

```diff
@@ -0,0 +1,142 @@
+package model
+
+import scala.util.Random
+
+/**
+ * Represents a Bunny.
+ */
+sealed trait Bunny{
+  val genotype: CompletedGenotype
+  val mom: Option[Bunny]
+  val dad: Option[Bunny]
+  val age: Int
+  val isAlive: Boolean
+
+  override def toString: String = {
+    super.toString + "\n" + genotype.genes
+      .map(g => "\t" + g._1 + ": "+ g._2.getVisibleTrait.toString.toLowerCase + " (" + g._2.getLetters + ")")
+      .reduce(_ + "\n" + _)
+      .replace("_", " ") + "\n"
+  }
+}
+
+/**
+ * Represents a Bunny which is still alive.
+ * @param genotype  contains the Bunny genes, it must be completed (contain all the available Genes of the world)
+ * @param mom       the bunny's mom
+ * @param dad       the bunny's dad
+ * @param age       how many generations the Bunny has spent in the world.
+ */
+case class AliveBunny(genotype: CompletedGenotype, mom:Option[Bunny], dad:Option[Bunny], age:Int) extends Bunny{
+  override val isAlive: Boolean = true
+}
+
+/**
+ * Represents a Bunny which dead.
+ * @param genotype  contains the Bunny genes, it must be completed (contain all the available Genes of the world)
+ * @param mom       the bunny's mom
+ * @param dad       the bunny's dad
+ * @param age       how many generations the Bunny has spent in the world.
+ */
+case class DeadBunny(genotype: CompletedGenotype, mom:Option[Bunny], dad:Option[Bunny], age:Int) extends Bunny{
```

Comment:
Ci serve una classe per rappresentare un Bunny morto? Perchè non basta mantenere aggiornato il campo isAlive?

Reply:
Perchè ho pensato che il Bunny avesse tutti i campi immutabili (val e non var) in modo che poi in ogni generazione ci sia una sorta di "snapshot" immutabile di tutti i conigli presenti in quella generazione. Essendo isAlive per "cambiarla" si deve direttamente creare un nuovo Bunny.

Reply:
Ho capito l'idea ma ho dei dubbi sulla sua funzionalità, ad esempio nel caso del metodo getNextGenerationBunnies come implementeresti il fatto di dover far morire i bunny che hanno raggiunto il numero massimo di generazioni?(morte di vecchiaia)

Reply:
Effettivamente in quel metodo pensavo solo ai nuovi conigli e ho dimenticato di inserire considerazioni su età/morte (come anche la costante di età massima). Pensandoci: per ogni Bunny si dovrebbe creare una nuova istanza con età diversa, se l'età raggiunge il massimo invece di metterlo nella lista di conigli si mette un'istanza si Bunny morto in un'apposita lista con tutti i morti. Una volta che il Bunny è morto non c'è bisogno di creare una nuova istanza per lui, quindi per ogni Bunny le istanze non crescono all'infinito ma ce ne sono maxAge+1. In ogni caso si dovrebbe avere un modo per legare fra loro le varie istanze di un Bunny nel tempo, una sorta di identificativo (ex. semplice numero crescente).

Reply:
C'è una cosa che non stiamo considerando che mi è venuta in mente prima: Cosa succede quando ci sono i lupi? Teoricamente dei conigli muoiono durante lo svolgimento della generazione, penso sia doppiamente costoso ricreare tutta la lista di bunnies in questa occorrenza

Reply:
Sono d'accordo che forse la lista dei bunny morti potrebbe essere molto onerosa e forse potrebbe anche essere in più del necessario. Io l'ho ragionata così: se invece di essere una variabile come campo di Bunny fosse una funzione _isAlive_? in questo modo ogni volta che c'è una generazione per visualizzare i conigli dovremo capire se ogni coniglio è vivo o morto, quindi check dell'età e se sono passati i lupi o meno. In questo modo ad ogni step della generazione possiamo sapere realtime se un coniglio è vivo o morto. Nel caso poi un coniglio muoia nella generazione successiva non dovrebbe essere più presente quindi potremo poi toglierlo da quella lista di coniglietti vivi (anche se così ho il dubbio che perderemo tutte le info di quel coniglio e potrebbe diventare un problema per i grafici forse)

Reply:
@anitvam nel caso dei lupi, la lista va ricreata solo le consideriamo immutabile (questo punto ancora non l'abbiamo toccato), per come la vedo io per ogni Bunny che era vivo si crea la sua versione "morta", quindi per quello specifico Bunny non avremo maxAge+1 versioni ma una in meno. @alessiarocco98 mi viene da pensare che non basta guardare se sono passati i lupi o meno per capire se è morto, ogni fattore deve avere la possibiltà di uccidere una certa % di coniglietti e dichiarare esattamente quali sono morti e quali vivi
Adesso l'ho implementate come dicevo sopra ma non mi convince, e mi è venuta un'altra idea: possiamo considerare un Bunny come solo il suo genotipo + madre e padre, gli aspetti che realmente sono immutabili. In questo modo possiamo avere una singola istanza immutabile per ogni Bunny, senza varie versioni. Poi, a livello di generazione associamo a quell'istanza di Bunny  un'oggettino molto meno pesante con solo l'age e l'isAlive, ovvero gli aspetti del Bunny che cambiano di generazione in generazione. Che ne pensate?